Lib/collections/__init__.py

```diff
@@ -920,7 +920,10 @@ def __len__(self):
         return len(set().union(*self.maps))     # reuses stored hash values if possible
 
     def __iter__(self):
-        return iter(set().union(*self.maps))
+        d = {}
+        for mapping in reversed(self.maps):
```

Comment:
Sorry, I was wrong. `reversed()` is not needed here.

Reply:
I think you were right and that the deepest mappings should determine the order while the shallowest mappings determine the value.  

This will be the most useful behavior in chains like command_line_args -> enviroment_variables -> default_values.  The default values can have the standard ordering while the overriding environment variables and command line arguments can be arbitrarily ordered.

Also, using reversed() is consistent with the notion of a ChainMap being like a base dictionary will several successive updates where the base layer determines order and the new layers either add new items or update values without changing their order.

Reply:
Currently `list(ChainMap({1: int}, {1.0: float}))` returns `[1]`. With `reversed()` it will return `[1.0]`.

Reply:
I think the [1.0] is correct.  It also matches what OrderedDict does when it encounters items with equal keys and different values:

    >>> OrderedDict([(1.0, float), (1, int)]).items() 
    odict_items([(1.0, <class 'int'>)])

Reply:
FWIW, except in the odd case of 1 and 1.0, the current behavior is unordered.  As an example, `list(ChainMap({'one': int}, {'two': float})` sometimes returns `['one']` and sometimes returns `['two']`.

Reply:
I understand your arguments. Such interpretation makes sense to me. But there are two changes in this PR:

1. The order was undetermined, now it will be determined. This doesn't break any code, because you could get the new order with old behavior.

2. In case of equal keys the behavior is changed. The first wined before, the last key will win with this change. This is subtle behavior change, but it is a change.

It would be better to discuss this on the tracker. This is not some implementation detail, but a design question.

Reply:
To be clear, this behavior change LGTM, but I just want to make clear that *there is* a behavior change.

Reply:
Yes, but I don't care about that behavior change.  The order was not previously guaranteed.  You're focusing on an odd and incidental behavior that was never "designed in" in the first place.

Are you opposed to going forward with the patch or do you prefer the current arbitrary and useless behavior?